### PR TITLE
Initial FRAM I2C

### PR DESCRIPTION
**Pull request by sommersoft**
Here it be. Fixes Part 1 of [CircuitPython Issue #665](https://github.com/adafruit/circuitpython/issues/665)

I assume RTD sub-project will need to be created, but that can wait until I finish the SPI version (separate PR).

Also, I'm going to put in an issue to update the FRAM Learn Guide with CircuitPython. I'm happy to do it, just need to be added to the guide.

### CODE REVIEW

**Comment by sommersoft** [comment]
Annnddd, Travis needs to be setup on this repo.

Also...needs librarians added to admins. (will this guy ever stop talking? 😆)

**Comment by kattni** [comment]
Added Librarians to the repo. Activated the repo on Travis, however it's not seeing the PR. I'll look more into it tomorrow.

File: adafruit_fram.py
Comment (caternuson) [inline adafruit_fram.py]: Could replace this entire function with:
```python
return self._wp if self._wp_pin is None else self._wp_pin.value
```

File: adafruit_fram.py
Comment (caternuson) [inline adafruit_fram.py]: Are there scenarios where this would need to be set or changed after creating the object? If not, remove this and require passing it in through `__init__`.

File: adafruit_fram.py
Comment (caternuson) [inline adafruit_fram.py]: Same comment here about pin object vs. designation.

File: adafruit_fram.py
Comment (caternuson) [inline adafruit_fram.py]: Use `__setitem__` for this to allow for `fram[address] = value`.
```python
fram[0x23] = 0x42
```

File: adafruit_fram.py
Comment (caternuson) [inline adafruit_fram.py]: This can also be done through `__setitem__`.
```python
fram[start_address] = data
```

File: adafruit_fram.py
Comment (caternuson) [inline adafruit_fram.py]: I2C bus creation should be done by the calling program and passed in. See how other I2C device drivers are doing this.
The MID/PID check to see if the device is connected and found is good though. Just change to doing that through the I2CDevice.

File: examples/fram_i2c_simpletest.py
Comment (caternuson) [inline examples/fram_i2c_simpletest.py]: Off by one. `fram.read(0)[0]`
But this will become
```python
print(fram[0])
```

**Review by caternuson — CHANGES_REQUESTED** [review]
General approach looks good. See comments about changes to interface.

**Comment by caternuson** [comment]
Just documenting my testing of the basic functionality of current version.
```python
Adafruit CircuitPython 3.0.3 on 2018-10-10; Adafruit ItsyBitsy M4 Express with samd51g19
>>> import board
>>> import adafruit_fram
>>> fram = adafruit_fram.FRAM_I2C(board.SCL, board.SDA)
>>> fram.max_size
32767
>>> fram.write_single(0, 1)
>>> print(fram.read(0)[0])
1
>>> fram.write_protected
False
>>> fram.write_protected = True
>>> fram.write_protected
True
>>> fram.write_single(1, 23)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "adafruit_fram.py", line 165, in write_single
RuntimeError: FRAM currently write protected.
>>> 
```

File: adafruit_fram.py
Comment (sommersoft) [reply adafruit_fram.py]: I can change to, and agree with, having the I2C bus passed in. However, I'll still need to use the raw bus to get the MID/PID due to the following:
- The Device ID (MID/PID) is at specific address, which is different from the I2C address space that is passed in.
- Reading the Device ID requires a certain order of writes/reads, which I2CDevice would not allow. When I2CDevice is [initialized](https://github.com/adafruit/Adafruit_CircuitPython_BusDevice/blob/master/adafruit_bus_device/i2c_device.py#L64), it writes an empty buffer to the stored address, which would (did) interfere with the specified steps.

File: adafruit_fram.py
Comment (sommersoft) [reply adafruit_fram.py]: I originally wrote it where the pin was only passed through during `__init__`. I added this later, as a "nicety" for the user. No specific scenario in mind... I am fine with removing it.

File: adafruit_fram.py
Comment (caternuson) [inline adafruit_fram.py]: Use`__getitem__` for this.
```python
value = fram[0]
values = fram[23:42]
```

**Review by caternuson — CHANGES_REQUESTED** [review]

**Comment by sommersoft** [comment]
Example of slice handling. Its rudimentary...doesn't support negative values, but aligns with CPython `range` assignment other than that.
```python
>>> fram[::1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "./circuitpython_libs/Adafruit_CircuitPython_FRAM/adafruit_fram.py", line 128, in __getitem__
MemoryError: memory allocation failed, allocating 4096 bytes
>>> fram[:100]
bytearray(b'\x00\x01\x1elohello\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\
x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\
x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\
x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\
x00\x00\x00\x00\x00\x00\x00\x00\x00\x00')
>>> fram[1:10:10]
bytearray(b'\x01')
>>> fram[1]
bytearray(b'\x01')
>>> fram[1:10]
bytearray(b'\x01\x1elohello')
>>> fram[:32760]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "./circuitpython_libs/Adafruit_CircuitPython_FRAM/adafruit_fram.py", line 128, in __getitem__
MemoryError: memory allocation failed, allocating 4096 bytes
>>> fram[32760:]
bytearray(b'\x00\x00\x00\x00\x00\x00\x00')
>>> fram[0:1:1] = [0,1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "./circuitpython_libs/Adafruit_CircuitPython_FRAM/adafruit_fram.py", line 160, in __setitem__
ValueError: Slice steps are not allowed during write operations.
>>> fram[0:5] = [0,1,2,3,4]
>>> fram[0:5]
bytearray(b'\x00\x01\x02\x03\x04')
```

File: adafruit_fram.py
Comment (caternuson) [inline adafruit_fram.py]: Maybe have this be:
```python
return self._read_byte(key)[0]
```
so you just get back the value instead of a single element bytearray with the value.

File: adafruit_fram.py
Comment (caternuson) [inline adafruit_fram.py]: Looks like slice code? Trying to use `.start` on an `int` type.

File: adafruit_fram.py
Comment (caternuson) [inline adafruit_fram.py]: Same here - trying to use `.start` on an `int` type.

**Review by caternuson — CHANGES_REQUESTED** [review]
New item access looks good. Found a few more tweaks.

File: adafruit_fram.py
Comment (sommersoft) [reply adafruit_fram.py]: hmm...swear I fixed that on the `fix the fixes` commit.

File: adafruit_fram.py
Comment (sommersoft) [reply adafruit_fram.py]: Meant to comment on this last night. With the change to `__getitem__`, I wanted to make the return value standard. I'm not completely opposed to sending back the value. Just felt that reduced API confusion was more beneficial.

File: adafruit_fram.py
Comment (caternuson) [inline adafruit_fram.py:209]: Do you have this here so it only gets imported when the I2C class is used?

Comment (sommersoft) [reply adafruit_fram.py:209]: Correct.

Comment (caternuson) [reply adafruit_fram.py:209]: OK. We may end up refactoring the various classes into other files. It's been done in other libraries, but I don't think we're being too consistent about it. This will become more important as this grows with the addition of the code for the SPI class.

**Review by caternuson — APPROVED** [review]

File: adafruit_fram.py
Comment (tannewt) [inline adafruit_fram.py]: Is it possible to read more than one byte at a time? That will be much faster.

File: adafruit_fram.py
Comment (tannewt) [inline adafruit_fram.py]: Should this be the standard `__len__` instead? That way you can use range on it in a loop.

**Review by tannewt — COMMENTED** [review]
A couple comments but looks good otherwise!

File: adafruit_fram.py
Comment (sommersoft) [reply adafruit_fram.py]: At this point, I don't remember why I chose to read one byte at a time. May have been an attempt to minimize memory use. 

Just ran some comparisons; memory allocation is tripped at the same length of reads both ways. I'll switch it. (it is much faster)

File: adafruit_fram.py
Comment (sommersoft) [reply adafruit_fram.py]: I have no issue with changing to `__len__`, if that is more Pythonic. Since `__getitem__` and `__setitem__` are already documented as special members, I can just add this too.

**Comment by sommersoft** [comment]
Don't merge yet...

I need to fix `__setitem__`; slicing touchup and handling `len(value) > slice[start:stop]`
```
>>> fram[0:4] = bytearray(b'length>greater_than_slice')
>>> fram[0:4]
bytearray(b'leng')
>>> fram[0:15]
bytearray(b'length>greater_')
```

**Comment by tannewt** [comment]
@sommersoft Ping me when you need another review.

**Comment by sommersoft** [comment]
@tannewt ready for another look. 

note: dropped slicing altogether for `__setitem__`, and removed slice stepping for `__getitem__`.

**Comment by kattni** [comment]
@tannewt ping.

**Review by tannewt — APPROVED** [review]
Looks great! Thanks!

**Pull request merged**
